### PR TITLE
Validator Assignment RPC Beacon Node Stream

### PR DESCRIPTION
**Pull request by rauljordan**
This is part of #545.

---

# Description

This PR implements the server-side functionality for streaming validator assignments to validator clients. Validators will send a request to a beacon node to watch a certain subset of public keys and the server will reply with a stream that is written to exactly at every cycle transition.

## Important Changes

This PR implements the following workflow:

1. Validator client connects to beacon node, requests genesis timestamp and list of assignments for the current cycle based on public keys passed in as a gRPC request
2. Validator client waits until current time is a multiple of params.SlotDuration
3. Validator client starts an internal ticker that checks if the current slot is its assigned slot and responsibility
4. Validator client listens to new assignments streamed by the beacon node every cycle transition

## Scope

The scope of this PR affects the RPC service in our beacon node, the beacon service in the validator clien, and their associated tests. This PR also implements a precise slot alignment ticker to ensure a beacon node streams to validator clients at exactly each slot interval. That is, if the current beacon slot started at 12:00:00PM and the current time is 12:00:03PM, we don't want to tick right away, but instead wait until 12:00:08PM to send a stream to connected validator clients.

### CODE REVIEW

File: beacon-chain/utils/clock.go [hunk 1]
```diff
@@ -1,9 +1,13 @@
 package utils
 
 import (
+	"math"
 	"time"
 )
 
+// GenesisTime used by the protocol.
+var GenesisTime = time.Date(2018, 9, 0, 0, 0, 0, 0, time.UTC) // September 2018
```
Comment (terenc3t) [inline beacon-chain/utils/clock.go]: I think GenesisTime should be a constant under `config.go`
https://notes.ethereum.org/SCIg8AH5SA-O4C1G1LYZHQ?view#Constants
What do you think?

Comment (rauljordan) [reply beacon-chain/utils/clock.go]: yeah, that's probably the best place to have it if someone wants to find it

File: beacon-chain/rpc/service.go [hunk 2]
```diff
@@ -238,7 +243,91 @@ func (s *Service) LatestCrystallizedState(req *empty.Empty, stream pb.BeaconServ
...
+				var role pb.ValidatorRole
+				if responsibility == "proposer" {
+					role = pb.ValidatorRole_PROPOSER
+				} else {
+					role = pb.ValidatorRole_ATTESTER
+				}
+
+				// We determine the assigned shard ID for the validator
+				// based on a public key and current crystallized state.
+				shardID, err := casper.ValidatorShardID(
```
Comment (terenc3t) [inline beacon-chain/rpc/service.go]: Another thought, should we just combine `ValidatorShardID` with `ValidatorSlotAndResponsibility`? That way ValidatorSlotAndResponsibility returns `slot, shard id and responsibility`?

Comment (terenc3t) [reply beacon-chain/rpc/service.go]: I can open a PR for you on that if you want

File: beacon-chain/utils/clock.go [hunk 3]
```diff
@@ -16,3 +19,19 @@ type RealClock struct{}
...
 }
+
+// CurrentBeaconSlot based on the seconds since genesis.
+func CurrentBeaconSlot() uint64 {
+	secondsSinceGenesis := time.Since(params.GenesisTime).Seconds()
+	return uint64(math.Floor(secondsSinceGenesis / 8.0))
+}
+
+// WaitUntilTimestamp sleeps until a specific time is reached after
+// a certain duratio. For example, if the genesis block
```
Comment (terenc3t) [inline beacon-chain/utils/clock.go]: s/duratio/duration

File: beacon-chain/blockchain/service.go [hunk 4]
```diff
@@ -63,14 +65,25 @@ func NewChainService(ctx context.Context, cfg *Config) (*ChainService, error) {
...
+
+	// If the genesis time was at 12:00:00PM and the current time is 12:00:03PM,
+	// the next slot should tick at 12:00:08PM. We can accomplish this
+	// using utils.WaitUntilTimestamp and passing in the desired
+	// slot duration.
+	//
+	// Instead of utilizing params.SlotDuration, we utilize a property of
+	// RPC service struct so this value can be set to 0 seconds
+	// as a parameter in tests. Otherwise, tests would sleep.
+	utils.WaitUntilTimestamp(time.Duration(c.slotAlignmentDuration) * time.Second)
```
Comment (terenc3t) [inline beacon-chain/blockchain/service.go]: I don't quite get this... Do we always wait 8 seconds before we start the NewTicker? And should this be inside `updateHead` similar to `ValidatorAssignment`?

Comment (rauljordan) [reply beacon-chain/blockchain/service.go]: We don't wait 8 seconds, we wait until the timestamp is a multiple of 8 seconds. So if the current slot started at 12:00:00 and the current time is 12:00:03, we wait until 12:00:08 to start ticking.

Comment (terenc3t) [reply beacon-chain/blockchain/service.go]: Got it!

Comment (prestonvanloon) [reply beacon-chain/blockchain/service.go]: Can this be `utils.BlockingWait`? There is no "timestamp" to speak of and the only parameter is a time.Duration.

File: beacon-chain/rpc/service.go [hunk 5]
```diff
@@ -212,6 +218,28 @@ func (s *Service) AttestHead(ctx context.Context, req *pb.AttestRequest) (*pb.At
...
+	for {
+		select {
+		case attestation := <-s.incomingAttestation:
+			log.Info("Sending attestation to RPC clients")
+			if err := stream.Send(attestation); err != nil {
+				return err
+			}
+		case <-sub.Err():
+			log.Debug("Subscriber closed, exiting goroutine")
+			return nil
```
Comment (prestonvanloon) [inline beacon-chain/rpc/service.go]: Can you return an error to the client?

File: beacon-chain/rpc/service.go [hunk 6]
```diff
@@ -212,6 +218,28 @@ func (s *Service) AttestHead(ctx context.Context, req *pb.AttestRequest) (*pb.At
...
+		case <-s.ctx.Done():
+			log.Debug("RPC context closed, exiting goroutine")
+			return nil
```
Comment (prestonvanloon) [inline beacon-chain/rpc/service.go]: Please return a "server closed stream" error to the client. This branch is likely to occur when the server shuts down so it would be helpful for the client to understand that the stream was close unexpectedly.

File: beacon-chain/rpc/service.go [hunk 7]
```diff
@@ -301,24 +323,107 @@ func (s *Service) ValidatorIndex(ctx context.Context, req *pb.PublicKey) (*pb.In
...
 }
 
-// LatestAttestation streams the latest processed attestations to the rpc clients.
-func (s *Service) LatestAttestation(req *empty.Empty, stream pb.BeaconService_LatestAttestationServer) error {
-	sub := s.attestationService.IncomingAttestationFeed().Subscribe(s.incomingAttestation)
-	defer sub.Unsubscribe()
+// ValidatorAssignment streams validator assignments every slot to clients that request
+// to watch a subset of public keys in the CrystallizedState's active validator set.
+func (s *Service) ValidatorAssignment(req *pb.ValidatorAssignmentRequest, stream pb.ValidatorService_ValidatorAssignmentServer) error {
+	// If the genesis time was at 12:00:00PM a if the SlotDuration is set to
```
Comment (prestonvanloon) [inline beacon-chain/rpc/service.go]: This comment seems to be repeated several times. Shouldn't it be part of the godoc and the interested reader can look there for more info?

File: beacon-chain/rpc/service.go [hunk 8]
```diff
@@ -301,24 +323,107 @@ func (s *Service) ValidatorIndex(ctx context.Context, req *pb.PublicKey) (*pb.In
...
+			log.Debug("ValidatorAssignment stream closed, exiting goroutine")
+			return nil
+		case <-tickerChan:
+
+			if len(req.GetPublicKeys()) == 0 {
+				return errors.New("no public keys specified in request")
+			}
+
+			log.WithField("slotNumber", currentSlot).Debug("Sending validator assignment to RPC clients")
+			// NOTE: this set will not change, so we can just cache the previous one
```
Comment (prestonvanloon) [inline beacon-chain/rpc/service.go]: ??

File: beacon-chain/rpc/service.go [hunk 9]
```diff
@@ -301,24 +323,107 @@ func (s *Service) ValidatorIndex(ctx context.Context, req *pb.PublicKey) (*pb.In
...
+
+			// We get the current crystallized state.
+			cState := s.chainService.CurrentCrystallizedState()
+
+			// Next, for each public key in the request, we build
+			// up an array of assignments.
+			assignments := []*pb.ValidatorAssignmentResponse_Assignment{}
+
+			for _, val := range req.GetPublicKeys() {
```
Comment (prestonvanloon) [inline beacon-chain/rpc/service.go]: This doesn't handle the case where the client requested all assignment updates.

Can you add that logic?

File: beacon-chain/rpc/service.go [hunk 10]
```diff
@@ -301,24 +323,107 @@ func (s *Service) ValidatorIndex(ctx context.Context, req *pb.PublicKey) (*pb.In
...
+
+				// For the corresponding public key and current crystallized state,
+				// we determine the assigned slot for the validator and whether it
+				// should act as a proposer or attester.
+				assignedSlot, responsibility, err := casper.ValidatorSlotAndResponsibility(
```
Comment (prestonvanloon) [inline beacon-chain/rpc/service.go]: What about when the pubkey is not assigned in this slot?

File: beacon-chain/rpc/service.go [hunk 11]
```diff
@@ -301,24 +323,107 @@ func (s *Service) ValidatorIndex(ctx context.Context, req *pb.PublicKey) (*pb.In
...
+	// Instead of utilizing params.SlotDuration, we utilize a property of
+	// RPC service struct called slotAlignmentDuration. This value
+	// can be set to 0 seconds as a parameter in tests.
+	// Otherwise, tests would sleep.
+	utils.WaitUntilTimestamp(s.slotAlignmentDuration)
+
+	// Right after the timestamps are aligned, we start a new ticker
+	// that fires exactly every params.SlotDuration*time.Second. This ensures
+	// a consistent "heartbeat" to keep track of every beacon slot.
+	tickerChan := time.NewTicker(s.slotAlignmentDuration).C
```
Comment (prestonvanloon) [inline beacon-chain/rpc/service.go]: Do we want any lead time? 

Shouldn't a validator know that they are assigned to a slot some amount of time prior to the slot so they can connect to peers on the assigned shard?

File: beacon-chain/utils/clock.go [hunk 12]
```diff
@@ -16,3 +19,19 @@ type RealClock struct{}
...
+	secondsSinceGenesis := time.Since(params.GenesisTime).Seconds()
+	return uint64(math.Floor(secondsSinceGenesis / 8.0))
+}
+
+// WaitUntilTimestamp sleeps until a specific time is reached after
+// a certain duration. For example, if the genesis block
+// was at 12:00:00PM and the current time is 12:00:03PM,
+// we want the next slot to tick at 12:00:08PM so we can use
+// this helper method to achieve that purpose.
+func WaitUntilTimestamp(duration time.Duration) {
```
Comment (prestonvanloon) [inline beacon-chain/utils/clock.go]: I don't quite like the name of this method, but I can't think of a better alternative.

File: proto/beacon/rpc/v1/services.proto [hunk 13]
```diff
@@ -67,12 +67,13 @@ message ValidatorAssignmentRequest {
 
 message ValidatorAssignmentResponse {
     repeated Assignment assignments = 1;
-    uint64 slot = 2;
+    uint64 current_beacon_slot = 2;
```
Comment (prestonvanloon) [inline proto/beacon/rpc/v1/services.proto]: Why change this?

The validator reasonably knows what slot they are a currently in, and this value serves as the slot they are expected to be assigned.

Please revert this change, unless there is an important reason for this change.

File: beacon-chain/rpc/service.go (hunk 11 above)
Comment (rauljordan) [reply beacon-chain/rpc/service.go]: This can be handled on the validator side. The validator will constantly listen for updates every slot from this stream and can then determine its role would happen N slots from the current one through that.

File: beacon-chain/rpc/service.go [hunk 14]
```diff
@@ -322,3 +369,104 @@ func (s *Service) LatestAttestation(req *empty.Empty, stream pb.BeaconService_La
...
+			Role:         role,
+			AssignedSlot: assignedSlot,
+		})
+	}
+	return assignments, nil
+}
+
+// allAssignmentsForCycle fetches the validator assignments for all active validators in
+// a given crystallized state's cycle.
+func allAssignmentsForCycle(cState *types.CrystallizedState) ([]*pb.Assignment, error) {
```
Comment (prestonvanloon) [inline beacon-chain/rpc/service.go]: This method is untested

File: beacon-chain/rpc/service.go [hunk 15]
```diff
@@ -158,16 +166,36 @@ func (s *Service) CanonicalHead(ctx context.Context, req *empty.Empty) (*pbp2p.B
...
-		LatestCrystallizedState: crystallized.Proto(),
+// CurrentAssignmentsAndGenesisTime returns the current validator assignments
+// based on the beacon node's current, canonical crystallized state.
+// Validator clients send this request once upon establishing a connection
+// to the beacon node in order to determine their role and assigned slot
+// initially. This method also returns the genesis timestamp
+// of the beacon node which will allow a validator client to setup a
+// a ticker to keep track of the current beacon slot.
+func (s *Service) CurrentAssignmentsAndGenesisTime(ctx context.Context, req *pb.ValidatorAssignmentRequest) (*pb.CurrentAssignmentsResponse, error) {
+	// #nosec G104
```
Comment (prestonvanloon) [inline beacon-chain/rpc/service.go]: Please provide a reason here.

Comment (rauljordan) [reply beacon-chain/rpc/service.go]: Done
